### DISCUSSION TITLE
Debugger Tools?

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#8380] [Usage Questions] [answered]
Hi, I'm using sqlalchemy with postgres. Both are amazing pieces of tech!

I was wondering if there are some good tools or ways to debug deadlock situations with multiple transactions hitting at once. 
I do turn on verbose logging and I can query the database for pgstats and connections, but looking for a better way to diagnose. Like what transactions the database has in the log and which are blocked, and why.

**Comment by zzzeek** [comment, marked as answer]
usually use pg_stat_activity, pg_locks: https://wiki.postgresql.org/wiki/Lock_Monitoring  should show you everything in play when a deadlock is occurring

**Reply by zzzeek** [reply]
https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-ACTIVITY-VIEW

**Reply by radiantone** [reply]
Thank you!!
